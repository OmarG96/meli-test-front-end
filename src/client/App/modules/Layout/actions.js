import {
  SEARCH_FETCH,
  SEARCH_FETCH_SUCCESS,
  SEARCH_FETCH_ERROR,
} from "./actionTypes";
import { searchItems } from "./services";

export const onFetchStart = () => ({
  type: SEARCH_FETCH,
});

export const onFetchSuccess = (data) => ({
  type: SEARCH_FETCH_SUCCESS,
  payload: { data },
});

export const onFetchError = (error) => ({
  type: SEARCH_FETCH_ERROR,
  payload: { error },
});

export const onFetch = (item) => async (dispatch) => {
  dispatch(onFetchStart());

  try {
    const data = await searchItems(item);
    dispatch(onFetchSuccess(data));
  } catch (error) {
    dispatch(onFetchError(error));
  }
};
