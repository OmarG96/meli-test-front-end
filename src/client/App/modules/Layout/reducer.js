import {
  SEARCH_FETCH,
  SEARCH_FETCH_SUCCESS,
  SEARCH_FETCH_ERROR,
} from "./actionTypes";

const initialState = {
  search: {
    loading: false,
    data: null,
    error: null,
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case SEARCH_FETCH:
      return {
        ...state,
        search: {
          data: null,
          loading: true,
          error: null,
        },
      };

    case SEARCH_FETCH_SUCCESS:
      const { data } = action.payload;
      return {
        ...state,
        search: {
          data,
          loading: false,
          error: null,
        },
      };

    case SEARCH_FETCH_ERROR:
      const { error } = action.payload;
      return {
        ...state,
        search: {
          ...state.search,
          loading: false,
          error: error,
        },
      };

    default: {
      return state;
    }
  }
};

export default reducer;
export { initialState };
