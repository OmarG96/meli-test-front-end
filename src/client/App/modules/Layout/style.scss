@import "../../../utils/scss/variables.scss";

.layout {
  background-color: $gray-1;
  min-height: 100vh;
}

.header-logo {
  height: 32px;
  margin-right: 24px;
  cursor: pointer;
}
