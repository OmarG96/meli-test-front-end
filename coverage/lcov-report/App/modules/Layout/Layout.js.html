<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for App/modules/Layout/Layout.js</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="..\..\..\prettify.css" />
    <link rel="stylesheet" href="..\..\..\base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(..\..\..\sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="..\..\..\index.html">All files</a> / <a href="index.html">App/modules/Layout</a> Layout.js
    </h1>
    <div class='clearfix'>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import React, { useCallback } from "react";
import { func, node } from "prop-types";
import { useHistory } from "react-router-dom";
&nbsp;
import { SearchBar, Header } from "../../../common";
&nbsp;
import logo from "../../../assets/images/Logo_ML.png";
import "./style.scss";
&nbsp;
const Layout = ({ children }) =&gt; {
  const history = useHistory();
&nbsp;
  const handleSearch = useCallback(
    (item) =&gt; {
      history.push(`/items?search=${item}`);
    },
    [history]
  );
&nbsp;
  const handleClickLogo = useCallback(() =&gt; {
    history.push("/");
  }, [history]);
&nbsp;
  return (
    &lt;div className="layout"&gt;
      &lt;Header&gt;
        &lt;img
          role="button-logo"
          className="header-logo"
          src={logo}
          onClick={handleClickLogo}
        /&gt;
        &lt;SearchBar onSearch={handleSearch} /&gt;
      &lt;/Header&gt;
      {children}
    &lt;/div&gt;
  );
};
&nbsp;
Layout.propTypes = {
  children: node,
  onSearch: func,
};
&nbsp;
export default Layout;
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Mon Jun 21 2021 18:32:13 GMT-0300 (GMT-03:00)
</div>
</div>
<script src="..\..\..\prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="..\..\..\sorter.js"></script>
</body>
</html>
