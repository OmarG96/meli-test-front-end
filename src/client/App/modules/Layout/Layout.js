import React, { useCallback } from "react";
import { func, node } from "prop-types";
import { useDispatch } from "react-redux";
import { useHistory } from "react-router-dom";

import { SearchBar, Header } from "../../../common";

import logo from "../../../assets/images/Logo_ML.png";
import "./style.scss";
import { onFetch } from "./actions";

const Layout = ({ children }) => {
  const dispatch = useDispatch();
  const history = useHistory();

  const handleSearch = useCallback(
    (item) => {
      dispatch(onFetch(item));
      history.push("/");
    },
    [dispatch, history]
  );

  const handleClickLogo = useCallback(() => {
    history.push("/");
  }, [history]);

  return (
    <div className="layout">
      <Header>
        <img
          role="button-logo"
          className="header-logo"
          src={logo}
          onClick={handleClickLogo}
        />
        <SearchBar onSearch={handleSearch} />
      </Header>
      {children}
    </div>
  );
};

Layout.propTypes = {
  children: node,
  onSearch: func,
};

export default Layout;
