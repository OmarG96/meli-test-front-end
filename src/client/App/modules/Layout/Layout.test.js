import React from "react";
import renderWithProvider from "../../../utils/tests/renderWithProvider";
import Layout from ".";
import { fireEvent, waitFor } from "@testing-library/react";

const mockPush = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    push: mockPush,
  }),
}));

describe("GIVEN Layout", () => {
  describe("WHEN is rendered", () => {
    it("THEN match to snapshot", async () => {
      const { asFragment } = renderWithProvider(
        <Layout>
          <div>Example</div>
        </Layout>
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it("THEN redirect to / when the logo is clicked", async () => {
      const { getByRole } = renderWithProvider(
        <Layout>
          <div>Example</div>
        </Layout>
      );
      await waitFor(() => fireEvent.click(getByRole("button-logo")));
      expect(mockPush).toHaveBeenCalledWith("/");
    });

    it("THEN should execute onFetch and redirect to / when a search is performed", async () => {
      const { getByRole } = renderWithProvider(
        <Layout>
          <div>Example</div>
        </Layout>
      );
      const searchbox = getByRole("searchbox");
      await waitFor(() => {
        fireEvent.change(searchbox, { target: { value: "example" } });
        fireEvent.submit(searchbox);
      });
      expect(mockPush).toHaveBeenCalledWith("/");
    });
  });
});
