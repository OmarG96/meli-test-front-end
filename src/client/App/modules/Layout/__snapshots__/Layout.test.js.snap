// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GIVEN Layout WHEN is rendered THEN match to snapshot 1`] = `
<DocumentFragment>
  <div
    class="layout"
  >
    <header>
      <div
        class="header-content"
      >
        <img
          class="header-logo"
          role="button-logo"
          src="test-file-stub"
        />
        <form
          class="search-bar"
        >
          <input
            class="search-input"
            name="search-input"
            placeholder="Nunca dejes de buscar"
            role="searchbox"
            value=""
          />
          <button
            class="search-button"
            type="submit"
          >
            <img
              src="test-file-stub"
            />
          </button>
        </form>
      </div>
    </header>
    <div>
      Example
    </div>
  </div>
</DocumentFragment>
`;
